vpr/src/route/rr_graph_uxsdcxx_serializer.h

```diff
@@ -1626,39 +1626,139 @@ class RrGraphSerializer final : public uxsd::RrGraphBase<RrGraphContextTypes> {
-            case uxsd::enum_loc_side::TOP:
-                return TOP;
             case uxsd::enum_loc_side::BOTTOM:
-                return BOTTOM;
+                side_mask.set(BOTTOM);
+                break;
+            case uxsd::enum_loc_side::BOTTOM_LEFT:
+                side_mask.set(BOTTOM);
+                side_mask.set(LEFT);
+                break;
+            case uxsd::enum_loc_side::LEFT:
+                side_mask.set(LEFT);
+                break;
+            case uxsd::enum_loc_side::TOP_RIGHT_BOTTOM_LEFT:
+                side_mask.set(TOP);
+                side_mask.set(RIGHT);
+                side_mask.set(BOTTOM);
+                side_mask.set(LEFT);
                 break;
             default:
                 report_error(
                     "Invalid side %d", side);
         }
-    }
-
-    uxsd::enum_loc_side to_uxsd_loc_side(e_side side) {
-        switch (side) {
-            case LEFT:
-                return uxsd::enum_loc_side::LEFT;
-            case RIGHT:
-                return uxsd::enum_loc_side::RIGHT;
-            case TOP:
-                return uxsd::enum_loc_side::TOP;
-            case BOTTOM:
-                return uxsd::enum_loc_side::BOTTOM;
-            default:
-                report_error(
-                    "Invalid side %d", side);
+        return side_mask;
+    }
+
+    /* A truth table to help understand the conversion from VPR side mask to uxsd side code
+     *
+     * index | LEFT BOTTOM RIGHT TOP | mask
+     * ======+======================+======
+     *  1    |                    X  | 0001
+     * ======+======================+======
+     *  2    |               X       | 0010
+     * ======+======================+======
+     *  3    |               X    X  | 0011
+     * ======+======================+======
+     *  4    |         X             | 0100
+     * ======+======================+======
+     *  5    |         X          X  | 0101
+     * ======+======================+======
+     *  6    |         X     X       | 0110
+     * ======+======================+======
+     *  7    |         X     X    X  | 0111
+     * ======+======================+======
+     *  8    |  X                    | 1000
+     * ======+======================+======
+     *  9    |  X                 X  | 1001
+     * ======+======================+======
+     *  10   |  X            X       | 1010
+     * ======+======================+======
+     *  11   |  X            X    X  | 1011
+     * ======+======================+======
+     *  12   |  X      X             | 1100
+     * ======+======================+======
+     *  13   |  X      X          X  | 1101
+     * ======+======================+======
+     *  14   |  X      X     X       | 1110
+     * ======+======================+======
+     *  15   |  X      X     X    X  | 1111
+     */
+    uxsd::enum_loc_side to_uxsd_loc_side(std::bitset<NUM_SIDES> sides) {
+        std::array<uxsd::enum_loc_side, 16> side_map;
```

Comment:
This should be initialized in the constructor for this class and re-used, rather than reconstructed everytime.  Other option is to make this a return from a constexpr function, but I don't know where the limit of constexpr functions and std::array.

Reply:
It makes sense. I know how to create a `constexpr` function like the `SIDES` variables. But I do not know if a constant side_map can be built using the same way. If you want, we can add the side_map as an internal data of the data structure and create it when initialize the object.

Reply:
> If you want, we can add the side_map as an internal data of the data structure and create it when initialize the object.

That was what I meant when I said "This should be initialized in the constructor for this class and re-used"